src/umap_.jl

```diff
@@ -75,42 +109,115 @@ function UMAP_(X::AbstractMatrix{S},
+- `local_connectivity::Integer = 1`: the number of nearest neighbors that should be assumed to be locally connected. The higher this value, the more connected the manifold becomes. This should not be set higher than the intrinsic dimension of the manifold.
+- `repulsion_strength::Real = 1`: the weighting of negative samples during the optimization process.
+- `neg_sample_rate::Integer = 5`: the number of negative samples to select for each positive sample. Higher values will increase computational cost but result in slightly more accuracy.
+- `a = nothing`: this controls the embedding. By default, this is determined automatically by `min_dist` and `spread`.
+- `b = nothing`: this controls the embedding. By default, this is determined automatically by `min_dist` and `spread`.
+"""
+function umap_transform(Q::AbstractMatrix{S},
+                        model::UMAP_;
+                        n_neighbors::Integer = 15,
+                        metric::Union{SemiMetric, Symbol} = Euclidean(),
+                        n_epochs::Integer = 100,
+                        learning_rate::Real = 1,
+                        min_dist::Real = 1//10,
+                        spread::Real = 1,
+                        set_operation_ratio::Real = 1,
+                        local_connectivity::Integer = 1,
+                        repulsion_strength::Real = 1,
+                        neg_sample_rate::Integer = 5,
+                        a::Union{Real, Nothing} = nothing,
+                        b::Union{Real, Nothing} = nothing
+                        ) where {S<:Real}
+    # argument checking
+    size(Q, 2) > n_neighbors > 0                     || throw(ArgumentError("size(Q, 2) must be greater than n_neighbors and n_neighbors must be greater than 0"))
+    learning_rate > 0                                || throw(ArgumentError("learning_rate must be greater than 0"))
+    min_dist > 0                                     || throw(ArgumentError("min_dist must be greater than 0"))
+    0 ≤ set_operation_ratio ≤ 1                      || throw(ArgumentError("set_operation_ratio must lie in [0, 1]"))
+    local_connectivity > 0                           || throw(ArgumentError("local_connectivity must be greater than 0"))
+    !isempty(model.data)                             || throw(ArgumentError("model.data must not be empty"))
+    size(model.data, 2) == size(model.embedding, 2)  || throw(ArgumentError("model.data must have same number of columns as model.embedding"))
+    size(model.data, 1) == size(Q, 1)                || throw(ArgumentError("size(model.data, 1) must equal size(Q, 1)"))
+
+
+    n_epochs = max(0, n_epochs)
+    # main algorithm
+    knns, dists = knn_search(model.data, Q, n_neighbors, metric, model.knns, model.dists)
```

Comment:
The verbosity would just be moved from one function to another, and this would additionally introduce some (logical) circular dependencies between "umap_.jl" (which defines the model) and "utils.jl" (which defines the `knn_search`, and has been so far loaded before "umap_.jl". The actual circular dependency could easily be hidden by removing any type constraints from the `knn_search` parameter and it would work perfectly fine but this just seems a bit more hacky to me... But overall this just seems like personal preference and I can still change this if you'd like.

Reply:
It's not a huge issue, so you can leave the arguments as they are for now.